src/tools/ceph_dedup_tool.cc

```diff
@@ -951,6 +956,246 @@ int chunk_scrub_common(const std::map < std::string, std::string > &opts,
+      } else if (fp_algo == "sha256") {
+        return ceph::crypto::digest<ceph::crypto::SHA256>(bl).to_str();
+      } else if (fp_algo == "sha512") {
+        return ceph::crypto::digest<ceph::crypto::SHA512>(bl).to_str();
+      } else {
+        assert(0 == "unrecognized fingerprint type");
+        return {};
+      }
+    }();
+    ret = chunk_io_ctx.write(chunk_object, bl, length, offset);
+    if (ret < 0) {
+      cerr << " writing object in chunk pool fails : " << cpp_strerror(ret) << std::endl;
+      goto out;
+    }
+    // 2. call set_chunk
+    ObjectReadOperation op;
+    op.set_chunk(offset, length, chunk_io_ctx, chunk_object, 0,
+	CEPH_OSD_OP_FLAG_WITH_REFERENCE);
+    ret = io_ctx.operate(object_name, &op, NULL);
+    if (ret < 0) {
+      cerr << " operate fail : " << cpp_strerror(ret) << std::endl;
+      goto out;
+    }
+  } else if (op_name == "object-dedup") {
+    unsigned chunk_size;
+    i = opts.find("dedup-cdc-chunk-size");
+    if (i != opts.end()) {
+      if (rados_sistrtoll(i, &chunk_size)) {
+	cerr << "unrecognized dedup_cdc_chunk_size " << chunk_size << std::endl;
+	return -EINVAL;
+      }
+    }
+
+    bufferlist inbl;
+    ret = rados.mon_command(
+	make_pool_str(pool_name, "fingerprint_algorithm", fp_algo),
+	inbl, NULL, NULL);
+    if (ret < 0) {
+      cerr << " operate fail : " << cpp_strerror(ret) << std::endl;
+      return ret;
+    }
+    ret = rados.mon_command(
+	make_pool_str(pool_name, "dedup_tier", chunk_pool_name),
+	inbl, NULL, NULL);
+    if (ret < 0) {
+      cerr << " operate fail : " << cpp_strerror(ret) << std::endl;
+      return ret;
+    }
+    ret = rados.mon_command(
+	make_pool_str(pool_name, "dedup_chunk_algorithm", "fastcdc"),
+	inbl, NULL, NULL);
+    if (ret < 0) {
+      cerr << " operate fail : " << cpp_strerror(ret) << std::endl;
+      return ret;
+    }
+    ret = rados.mon_command(
+	make_pool_str(pool_name, "dedup_cdc_chunk_size", chunk_size),
+	inbl, NULL, NULL);
+    if (ret < 0) {
+      cerr << " operate fail : " << cpp_strerror(ret) << std::endl;
+      return ret;
+    }
+
+    // convert object to manifest object
+    ObjectWriteOperation op;
+    op.create(true);
+
+    auto gen_r_num = [] () -> string {
+      std::random_device rd;
+      std::mt19937 gen(rd());
+      std::uniform_int_distribution<uint64_t> dist;
+      uint64_t r_num = dist(gen);
+      return to_string(r_num);
+    };
+    string temp_oid = gen_r_num();
+    // create temp chunk object for set-chunk
+    ret = chunk_io_ctx.operate(temp_oid, &op);
+    if (ret == -EEXIST) {
+      // one more try
+      temp_oid = gen_r_num();
+      ret = chunk_io_ctx.operate(temp_oid, &op);
+    } 
+    if (ret < 0) {
+      cerr << " operate fail : " << cpp_strerror(ret) << std::endl;
+      goto out;
+    }
+
+    // set-chunk to make manifest object
+    ObjectReadOperation chunk_op;
+    chunk_op.set_chunk(0, 4, chunk_io_ctx, temp_oid, 0,
```

Comment:
Why are we setting the first 4 bytes of the target object to point at an empty temp object?

Reply:
I added write_full(4 bytes) before this. The reason why set_chunk is needed here is that set_chunk is the only way to set the normal object to the manifest object.

Reply:
'temp' isn't necessarily the first 4 bytes of the object?  What happens if we set_chunk 4 bytes that don't actually match the object contents?  What if the object is smaller than 4 bytes?  I think I'm missing something -- this really doesn't make sense to me.

Reply:
The purpose of this set_chunk() is only to convert normal object to manifest object because tier_flush(), which will be called right after this set_chunk(), performs deduplication if the object is manifest object. 
See below.

1. call set_chunk
- object (base tier) ---set_chunk---> temp object (chunk tier)
2. the object is now manifest object
3. call tier_flush 
- object (base tier) ---tier_flush---> chunk objects (chunk tier, temp object is deleted)

At this point, tier_flush re-calculate all fingerprints based on the entire object's contents (not temp object), decreasing old references.
So, whatever arguments---4 bytes contents or something---set_chunk() use, the arguments are meaningless because the temp object will be deleted after tier_flush---reading object's current contents, then recalculating fingerprints---is called.
Note that temp object---which has temp_oid---is needed here because set_chunk() requires a target object that presents in the pool.

I think this pre-processing---calling set_chunk before tier_flush---might make user confused. We can avoid this step if we allow the object to be converted to manifest object when tier_flush is called.

Reply:
Using set_chunk on an incorrect chunk in order to make the object a manifest object isn't really ok.  Is there a disadvantage to tier_flush implicitly making the object a manifest object?

Reply:
Hmm, TIER_FLUSH behaves differently for a cache pool, implicitly converting to a manifest object may not make sense.

Reply:
Hmm.  We need a better solution here, but we can merge this with the hack for now.  Add a comment

```
/* TODO: add a better way to make an object a manifest object.  We're using set_chunk with an
 * incorrect object here simply to make the object a manifest object, the tier_flush() will remove
 * it and replace it with the real contents.
 */
```

Reply:
Does this hack actually work if the object is shorter than 4 bytes?

Reply:
> Using set_chunk on an incorrect chunk in order to make the object a manifest object isn't really ok. Is there a disadvantage to tier_flush implicitly making the object a manifest object?

In my opinion, these is no problem if tier_flush has only effect on the object which is configured by CDC and fingerprint options.

> Does this hack actually work if the object is shorter than 4 bytes?

Yes, except for 0 bytes.

Reply:
But, as you mentioned, this is unexpected behavior if current manifest object supports both cache tier (future work) and dedup---we need to classify non-dedup(for cache) and dedup object(for dedup).